### PR TITLE
fix: remove constraint before alter foreign keys

### PR DESCRIPTION
**Pull request by lext-7**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm run test` or `npm run test-DIALECT` pass with this change (including linting)?
- [ ] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Have you added new tests to prevent regressions?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Did you follow the commit message conventions explained in [CONTRIBUTING.md](../CONTRIBUTING.md)?

### Description of change

Remove constraint before alter foreign keys. When sync with alter, it will throw an error `Can't write; duplicate key in table '#sql-xxx'`  because there is an existing constraint at the foreign key. 

Every time altering models, it should get all foreign key informations and remove duplicated foreign key constraints, recreate them.

### CODE REVIEW

**Comment by robin-garnham** [comment]
Thanks lext-7... I'm hoping this gets merged soon... this is causing me pain daily...

**Comment by n8stowell82** [comment]
I very much need this fix as I am currently blocked from making upgrades to my models.  Thank you!

File: lib/dialects/mysql/query-generator.js [hunk 1]
```diff
@@ -495,15 +523,17 @@ const QueryGenerator = {
    * @private
    */
   getForeignKeyQuery(table, columnName) {
-    let tableName = table.tableName || table;
-    if (table.schema) {
-      tableName = table.schema + '.' + tableName;
-    }
-    return 'SELECT CONSTRAINT_NAME as constraint_name'
+    const tableName = table.tableName || table;
+
```
Comment (sushantdhiman) [inline lib/dialects/mysql/query-generator.js]: Remove this line

File: lib/dialects/mysql/query-interface.js [hunk 2]
```diff
@@ -26,7 +26,10 @@ function removeColumn(tableName, columnName, options) {
   options = options || {};
 
   return this.sequelize.query(
-    this.QueryGenerator.getForeignKeyQuery(tableName, columnName),
+    this.QueryGenerator.getForeignKeyQuery(tableName.tableName ? tableName : {
+      tableName,
+      schema: this.sequelize.config.database
```
Comment (sushantdhiman) [inline lib/dialects/mysql/query-interface.js]: This should be `config.schema`

File: lib/model.js [hunk 3]
```diff
@@ -1122,18 +1122,46 @@ class Model {
...
+                } else if (!currentAttributes.primaryKey) {
+                  // Check foreign keys. If it's a foreign key, it should remove constraint first.
+                  const references = currentAttributes.references;
+                  if (currentAttributes.references) {
+                    const database = this.sequelize.config.database;
+                    // Find existed foreign keys
+                    _.each(foreignKeyReferences, foreignKeyReference => {
+                      const constraintName = foreignKeyReference.constraintName;
+                      if (!!constraintName
+                        && foreignKeyReference.tableCatalog === database
```
Comment (sushantdhiman) [inline lib/model.js]: This should be checking for `schema`

File: lib/dialects/mysql/query-generator.js (hunk 1 above)
Comment (lext-7) [reply lib/dialects/mysql/query-generator.js]: I think it should keep the same as https://github.com/sequelize/sequelize/blob/master/lib/dialects/mssql/query-generator.js#L649

File: lib/dialects/mysql/query-interface.js (hunk 2 above)
Comment (lext-7) [reply lib/dialects/mysql/query-interface.js]: https://github.com/sequelize/sequelize/blob/master/lib/dialects/mysql/query-generator.js#L484
https://github.com/sequelize/sequelize/blob/master/lib/query-interface.js#L697
I think in mysql, schema can be seen as database. There is no schema option, no specific schema concept in sequelize, just use database for all.

File: lib/dialects/mysql/query-generator.js (hunk 1 above)
Comment (sushantdhiman) [reply lib/dialects/mysql/query-generator.js]: I meant the empty line between two const definition

Comment (lext-7) [reply lib/dialects/mysql/query-generator.js]: Sorry for misunderstanding

File: lib/dialects/mysql/query-interface.js (hunk 2 above)
Comment (sushantdhiman) [reply lib/dialects/mysql/query-interface.js]: Got it, can you add a comment here explaining why schema is being assigned database

**Pull request merged**